python/lance/types/image.py

```diff
@@ -0,0 +1,206 @@
+    def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        return ImageUriType()
+
+
+class ImageBinaryType(ImageType):
+    """
+    Represents an inlined Image containing the actual image bytes. Storage type
+    is binary.
+
+    TODO: add support for large binary
+    """
+
+    def __init__(self):
+        super(ImageBinaryType, self).__init__(pa.binary(), "image[binary]")
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        return ImageBinaryType()
+
+
+class Image(ABC):
+    """
+    Represents an Image that can either be in-memory or stored at some uri.
+    This class allows us to convert to/from Image bytes, uri, numpy, and PIL
+    representations
+    """
+
+    def __new__(cls, data: Union[bytes, str]):
+        if isinstance(data, bytes):
+            img = object.__new__(EmbeddedImage)
+        elif isinstance(data, str):
+            img = object.__new__(ImageRef)
+        else:
+            raise TypeError(
+                f"{cls.__name__} can only handle bytes or str " f"but got {type(data)}"
+            )
+        img.__init__(data)
+        return img
+
+    @classmethod
+    def from_dict(cls, data: dict):
+        """
+        Construct Image from result of Image.to_dict
+        """
+        return Image(data.get("data", data.get("uri")))
+
+    @abstractmethod
+    def open(self) -> IO:
+        """Open for reading"""
+        pass
+
+    def to_pil(self, mode="RGB") -> "PILImage":
+        """Get a PIL instance converted to RGB by default"""
+        return PILImage.open(self.open()).convert(mode=mode)
+
+    def to_numpy(self) -> np.ndarray:
+        """Get the 3d numpy array representation of this image"""
+        with self.to_pil() as pil_img:
+            return np.asarray(pil_img)
+
+    @abstractmethod
+    def to_dict(self) -> dict:
+        """Return a json friendly dict"""
+        pass
+
+    @abstractmethod
+    def to_embedded(self) -> "Image":
+        """Return an in-memory Image instance"""
+        pass
+
+    @abstractmethod
+    def save(self, uri: Union[str, Path]) -> "Image":
+        """Write this image to the given uri and return the new Image"""
+        pass
+
+
+class EmbeddedImage(Image):
+    """
+    In-memory Image
+    """
+
+    def __init__(self, data: bytes):
+        self._bytes = data
+
+    @property
+    def bytes(self) -> bytes:
+        return self._bytes
+
+    def open(self) -> IO:
+        return io.BytesIO(self.bytes)
+
+    def to_dict(self) -> dict:
+        return {"data": self.bytes}
+
+    def to_embedded(self):
+        return self
+
+    def save(self, uri):
+        filesystem, path = fs.FileSystem.from_uri(uri)
+        with filesystem.open_output_stream(path) as fobj:
+            fobj.write(self.bytes)
+        return ImageRef(uri)
+
+    def __eq__(self, other):
+        return isinstance(other, EmbeddedImage) and other.bytes == self.bytes
+
+
+class ImageRef(Image):
+    """
+    Image reference
```

Comment:
`ImageRef` is an image object with URL?

Can you add some comments to make it clear? Do users directly use this class or they mostly create this object via `Image`?